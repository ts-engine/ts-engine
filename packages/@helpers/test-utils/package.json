{
  "name": "@helpers/test-utils",
  "version": "1.1.1",
  "license": "MIT",
  "private": true,
  "main": "dist/main.cjs.js",
  "module": "dist/main.esm.js",
  "types": "dist/main.d.ts",
  "scripts": {
    "build": "rm -rf dist && ts-engine build --library",
    "lint": "ts-engine lint",
    "typecheck": "ts-engine typecheck"
  },
  "devDependencies": {
    "@ts-engine/cli": "1.1.1"
  },
  "dependencies": {
    "chalk": "^4.0.0",
    "fs-extra": "^9.0.0",
    "strip-ansi": "^6.0.0",
    "terminate": "^2.1.2"
  }
}
