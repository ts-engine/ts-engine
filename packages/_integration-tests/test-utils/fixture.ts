import path from "path";
import { spawnSync } from "child_process";
import fs from "fs-extra";
import { parseArgsStringToArgv } from "string-argv";
import stripAnsi from "strip-ansi";

interface RunCliResult {
  status: number;
  stdout: string;
  stderr: string;
}

interface RunToolOptions {
  name: string;
  cwd: string;
  args: string;
}

const runTool = (options: RunToolOptions) => {
  const result = spawnSync(options.name, parseArgsStringToArgv(options.args), {
    cwd: options.cwd,
    encoding: "utf8",
  });

  return {
    status: result.status,
    stdout: stripAnsi(result.stdout),
    stderr: stripAnsi(result.stderr),
  };
};

interface Fixture {
  reset: () => Promise<void>;
  runTse: (args: string) => RunCliResult;
  runNode: (args: string) => RunCliResult;
  writeFile: (filepath: string, content: string) => Promise<void>;
  removeFile: (filepath: string) => Promise<void>;
  readFile: (filepath: string) => Promise<string>;
}

type FixtureName =
  | "build-normal"
  | "build-minify"
  | "build-react"
  | "build-syntax-error"
  | "build-type-error"
  | "lint-error"
  | "lint-fix"
  | "lint-pass"
  | "lint-warning"
  | "lint-no-files"
  | "run-normal"
  | "run-minify"
  | "run-react"
  | "run-syntax-error"
  | "run-type-error"
  | "test-error"
  | "test-pass"
  | "test-config"
  | "test-setup-ts"
  | "test-setup-js"
  | "test-no-tests"
  | "version";

export const createFixture = (name: FixtureName): Fixture => {
  const dir = path.resolve("../_integration-fixtures/", name);
  const distDir = path.resolve(dir, "dist");

  const reset = async () => {
    await fs.remove(distDir);
  };

  const runTse = (args: string) => {
    return runTool({
      name: "yarn",
      cwd: dir,
      args: `run tse ${args}`,
    });
  };

  const runNode = (args: string) => {
    return runTool({
      name: "node",
      cwd: dir,
      args,
    });
  };

  const writeFile = async (filepath: string, content: string) => {
    await fs.writeFile(path.resolve(dir, filepath), content, {
      encoding: "utf8",
    });
  };

  const removeFile = async (filepath: string) => {
    await fs.remove(path.resolve(dir, filepath));
  };

  const readFile = async (filepath: string) => {
    return await fs.readFile(path.resolve(dir, filepath), { encoding: "utf8" });
  };

  return {
    reset,
    runTse,
    runNode,
    writeFile,
    removeFile,
    readFile,
  };
};
