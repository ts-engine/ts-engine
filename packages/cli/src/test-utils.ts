import { parseArgsStringToArgv } from "string-argv";
import { cli } from "./cli";

export const testConsole = {
  log: console.debug,
};

interface RunCliResult {
  stdout: string[];
  stderr: string[];
  exitCode: number;
}

export const runCli = async (args: string): Promise<RunCliResult> => {
  // capture stdout/stderr console logs
  const stderr: string[] = [];
  const stdout: string[] = [];

  const stderrSpy = jest.fn((message: any) => {
    if (message) {
      stderr.push(message.toString());
    }
  });

  const stdoutSpy = jest.fn((message: any) => {
    if (message) {
      stdout.push(message.toString());
    }
  });

  jest.spyOn(console, "error").mockImplementation(stderrSpy);
  jest.spyOn(console, "info").mockImplementation(stdoutSpy);
  jest.spyOn(console, "log").mockImplementation(stdoutSpy);
  jest.spyOn(console, "warn").mockImplementation(stdoutSpy);

  // capture exit code
  let exitCode = 0;
  const exitPromise = new Promise((resolve) => {
    // @ts-ignore
    jest.spyOn(process, "exit").mockImplementation((code: number) => {
      // only assign if the code is still 0 (falsy) as
      // process.exit can only be called once
      exitCode = exitCode || code;
      resolve();
    });
  });

  // run cli
  cli(parseArgsStringToArgv(args));

  // let async code complete
  await exitPromise;

  // reset mocks
  (console.error as jest.Mock).mockRestore();
  (console.info as jest.Mock).mockRestore();
  (console.log as jest.Mock).mockRestore();
  (console.warn as jest.Mock).mockRestore();

  // @ts-ignore
  (process.exit as jest.Mock).mockRestore();

  // collate results
  return {
    stderr,
    stdout,
    exitCode,
  };
};

export const matchLog = (regex: RegExp | string, logs: string[]) => {
  return logs.find((log) => log.match(regex));
};
