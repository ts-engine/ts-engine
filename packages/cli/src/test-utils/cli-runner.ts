import { cli } from "../cli";

interface RunCliResult {
  stdout: string[];
  stderr: string[];
  exitCode: number;
}

export const runCli = async (...args: string[]): Promise<RunCliResult> => {
  // capture stdout/stderr console logs
  const stderr: string[] = [];
  const stdout: string[] = [];

  const stderrSpy = jest.fn((message: any) => {
    stderr.push(message);
  });

  const stdoutSpy = jest.fn((message: any) => {
    stdout.push(message);
  });

  console.error = jest.fn(stderrSpy);
  console.info = jest.fn(stdoutSpy);
  console.log = jest.fn(stdoutSpy);
  console.warn = jest.fn(stdoutSpy);

  // capture exit code
  let exitCode = 0;

  // @ts-ignore
  process.exit = jest.fn((code: number) => (exitCode = code));

  // run cli
  cli(args);

  // let async code complete
  await new Promise((resolve) => setImmediate(resolve));

  // reset mocks
  (console.error as jest.Mock).mockRestore();
  (console.info as jest.Mock).mockRestore();
  (console.log as jest.Mock).mockRestore();
  (console.warn as jest.Mock).mockRestore();

  // @ts-ignore
  (process.exit as jest.Mock).mockRestore();

  // collate results
  return {
    stderr,
    stdout,
    exitCode,
  };
};
