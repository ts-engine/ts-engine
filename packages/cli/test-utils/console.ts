interface MockConsoleResult {
  log: jest.Mock;
  error: jest.Mock;
  warn: jest.Mock;
  reset: () => void;
}

export const mockConsole = (): MockConsoleResult => {
  const consoleLogMock = jest.fn();
  const consoleErrorMock = jest.fn();
  const consoleWarnMock = jest.fn();

  console.log = consoleLogMock;
  console.error = consoleErrorMock;
  console.warn = consoleWarnMock;

  return {
    log: consoleLogMock,
    error: consoleErrorMock,
    warn: consoleWarnMock,
    reset: () => {
      consoleLogMock.mockReset();
      consoleErrorMock.mockReset();
      consoleWarnMock.mockReset();
    },
  };
};
