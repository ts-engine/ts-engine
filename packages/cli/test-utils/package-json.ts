export const createPackageJsonMock = () => {
  return {
    name: "example",
    version: "1.0.0",
    description: "Example package.json",
  };
};
