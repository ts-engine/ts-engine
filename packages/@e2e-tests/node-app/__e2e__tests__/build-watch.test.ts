import { fileSystem, runCliCommand } from "@helpers/test-utils";

describe("build-watch", () => {
  let revertFileEdit: () => Promise<void>;

  beforeEach(async () => {
    await fileSystem.deleteDir("dist");
  });

  afterEach(async () => {
    await revertFileEdit();
  });

  it("should rebuild on changes", async () => {
    const runner = runCliCommand("yarn run ts-engine build --node-app --watch");

    // Wait for tool to start waiting for changes
    await runner.waitUntilStdoutLine("Watching for changes...");

    // Printed info to stdout
    expect(runner.stdoutLines).toContainInOrder([
      "Building code with Rollup",
      "src/main.ts ⮕  dist/main.js",
      "Watching for changes...",
    ]);

    // Built file is written to file system
    expect(await fileSystem.fileExists("dist/main.js")).toBe(true);

    // Clean up build
    await fileSystem.deleteDir("dist");

    // Update file
    revertFileEdit = await fileSystem.editFileTemporarily(
      "src/main.ts",
      `console.log("new content");`
    );

    // Wait for tool to start waiting for changes again
    await runner.waitUntilStdoutLine("Watching for changes...");

    // Printed info to stdout
    expect(runner.stdoutLines).toContainInOrder([
      "Building code with Rollup",
      "src/main.ts ⮕  dist/main.js",
      "Watching for changes...",
      "src/main.ts ⮕  dist/main.js",
      "Watching for changes...",
    ]);

    // Built file is written to file system
    expect(await fileSystem.readFile("dist/main.js")).toMatchSnapshot(
      "updated"
    );

    // Kill watching tool
    runner.kill();
  });
});
