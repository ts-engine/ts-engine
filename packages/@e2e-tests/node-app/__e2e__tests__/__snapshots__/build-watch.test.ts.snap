// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`build-watch should rebuild on changes: updated 1`] = `
"\\"use strict\\";console.log(\\"new content\\");
"
`;
